File: problem11/kotlin/MatrixProductCLI.kt
import java.io.File

fun readMatrix(filename: String): List<List<Int>> {
    return File(filename).readLines().filter { it.isNotBlank() }.map { line ->
        line.trim().split(Regex("\\s+")).map { it.toInt() }
    }
}

fun printMatrixWithHighlight(matrix: List<List<Int>>, coords: List<Pair<Int, Int>>) {
    val coordSet = coords.toSet()
    for ((i, row) in matrix.withIndex()) {
        for ((j, value) in row.withIndex()) {
            if (coordSet.contains(Pair(i, j))) {
                print("\u001B[31m%02d\u001B[0m ".format(value))
            } else {
                print("%02d ".format(value))
            }
        }
        println()
    }
}

fun main(args: Array<String>) {
    val filename = if (args.isNotEmpty()) args[0] else "matrix.txt"
    val matrix = try {
        readMatrix(filename)
    } catch (e: Exception) {
        System.err.println("Failed to read matrix from $filename: ${e.message}")
        return
    }
    val (max, coords) = greatestProduct(matrix, 4)
    printMatrixWithHighlight(matrix, coords)
    println("Greatest product of four adjacent numbers: $max")
}

File: problem11/ruby/matrix_product_cli.rb
require_relative 'matrix_product'

# Reads a matrix from a file where each line is a row of space-separated integers.
def read_matrix(filename)
  File.readlines(filename).map { |line| line.split.map(&:to_i) }
end

# Prints the matrix, highlighting the winning sequence in red.
def print_matrix_with_highlight(matrix, coords)
  coord_set = coords.map { |c| c }.to_set
  matrix.each_with_index do |row, i|
    row.each_with_index do |val, j|
      if coord_set.include?([i, j])
        print "\e[31m%02d\e[0m " % val
      else
        print "%02d " % val
      end
    end
    puts
  end
end

if __FILE__ == $0
  require 'set'
  filename = ARGV[0] || 'matrix.txt'
  begin
    matrix = read_matrix(filename)
  rescue => e
    warn "Failed to read matrix from #{filename}: #{e}"
    exit 1
  end
  result, coords = greatest_product(matrix, 4)
  print_matrix_with_highlight(matrix, coords)
  puts "Greatest product of four adjacent numbers: #{result}"
end
